### PR TITLE
Task/upgrades

### PR DESCRIPTION
**Pull request by m-juckes-pins**
# Pull Request Template

## Describe your changes

Starting with modules as other repos rely on it and you can't upgrade as there will be terraform constraints errors. 

Changed the providers to `version = ">= 3.0, < 5.0"` to include all versions to hopefully avoid any constraint issues..

- DEV 275

## Useful information to review or test

- Add any useful information that will help the reviewer test your changes.
- This could include example payloads, steps to reproduce, etc.

## Type of change 🧩

- [ ] Adding a new module
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Other (please explain in the description section above)

## Checklist before requesting a review

- [ ] My code follows the style guidelines of this project
- [ ] My code changes do not include any hardcoded secrets or passwords
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] My Tech Lead is aware of any infrastructure changes that will cost money

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -14,5 +14,3 @@ repos:
         args:
           - --args=--config=__GIT_WORKING_DIR__/.tflint.hcl
       - id: terraform_checkov
-        args:
-          - --args=--config-file=__GIT_WORKING_DIR__/.checkov.yaml
```
Comment (m-juckes-pins) [inline .pre-commit-config.yaml:18]: removing as we don't have a checkov file in a repo or pinned to a specific version. We need to do the major update for checkov...

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -1,18 +1,56 @@
...
+        exclude: ^examples
+        args:
+          - --tf-init-args=-upgrade
+          - --hook-config=--retry-once-with-cleanup=true
       - id: terraform_tflint
+        stages: [pre-commit]
+        exclude: ^examples
         args:
           - --args=--config=__GIT_WORKING_DIR__/.tflint.hcl
+      - id: terraform_trivy
```
Comment (m-juckes-pins) [inline .pre-commit-config.yaml:47]: checks for security vulnerabilities, secrets, misconfigurations

Comment (m-juckes-pins) [reply .pre-commit-config.yaml:47]: This will also require an install of trivy on the developer's machine

Comment (m-juckes-pins) [reply .pre-commit-config.yaml:47]: Can also be used for Azure CI/CD pipelines..

**Review by Medini-Aithal — DISMISSED** [review]

**Review by beejjacobs — COMMENTED** [review]
Looks good, I think the pre-commit changes need some documentation with them perhaps.

File: modules/node-function-app/test/variables.tf [hunk 3]
```diff
@@ -19,17 +19,17 @@ variable "app_service_plan_id" {
   type        = string
 }
```
Comment (m-juckes-pins) [inline modules/node-function-app/test/variables.tf:21]: again, i can delete these. Not sure why TF Validate was not picking all this up before

File: modules/node-app-service/variables.tf [hunk 4]
```diff
@@ -83,6 +83,12 @@ variable "health_check_path" {
   default     = null
 }
 
+variable "health_check_eviction_time_in_min" {
+  description = "The path of the service's health check endpoint"
+  type        = number
+  default     = 10
```
Comment (m-juckes-pins) [inline modules/node-app-service/variables.tf:89]: I've literally just put 10 in there now just to pass

**Comment by m-juckes-pins** [comment]
at the moment there is some left of JS stuff in here that I could delete, this was from trying to get commintlint working

File: modules/node-app-service/locals.tf [hunk 5]
```diff
@@ -2,10 +2,10 @@ locals {
...
-      DOCKER_REGISTRY_SERVER_PASSWORD = sensitive(data.azurerm_container_registry.acr.admin_password)
-      DOCKER_REGISTRY_SERVER_URL      = data.azurerm_container_registry.acr.login_server
-      DOCKER_REGISTRY_SERVER_USERNAME = data.azurerm_container_registry.acr.admin_username
+      DOCKER_ENABLE_CI         = true
+      docker_registry_password = sensitive(data.azurerm_container_registry.acr.admin_password)
+      docker_registry_url      = data.azurerm_container_registry.acr.login_server
+      docker_registry_username = data.azurerm_container_registry.acr.admin_username
     }
   )
 }
```
Comment (m-juckes-pins) [inline modules/node-app-service/locals.tf:11]: https://registry.terraform.io/providers/hashicorp/azurerm/4.5.0/docs/resources/linux_web_app
NOTE:
docker_registry_url, docker_registry_username, and docker_registry_password replace the use of the app_settings values of DOCKER_REGISTRY_SERVER_URL, DOCKER_REGISTRY_SERVER_USERNAME and DOCKER_REGISTRY_SERVER_PASSWORD respectively, these values will be managed by the provider and should not be specified in the app_settings map.

File: modules/node-app-service/monitoring.tf [hunk 6]
```diff
@@ -70,6 +70,7 @@ resource "azurerm_monitor_metric_alert" "app_service_response_time" {
 resource "azurerm_monitor_activity_log_alert" "app_service_stop" {
   name                = "App Service Stopped - ${reverse(split("/", azurerm_linux_web_app.web_app.id))[0]}"
   resource_group_name = var.resource_group_name
+  location            = var.location
```
[the remark this answers is not in the document]
Comment (m-juckes-pins) [reply modules/node-app-service/monitoring.tf]: Can be "Global"

File: modules/node-function-app/test/variables.tf (hunk 3 above)
Comment (m-juckes-pins) [reply modules/node-function-app/test/variables.tf:21]: Not a massive concern as in the test folder

File: modules/node-function-app/monitoring.tf [hunk 7]
```diff
@@ -1,130 +1,132 @@
...
+
+  action {
+    action_group_id = var.action_group_ids.tech
+  }
+}
+
+resource "azurerm_monitor_activity_log_alert" "function_app_stop" {
+  name                = "Function App Stopped - ${reverse(split("/", azurerm_linux_function_app.function_app.id))[0]}"
+  resource_group_name = var.resource_group_name
+  location            = "Global"
```
Comment (m-juckes-pins) [inline modules/node-function-app/monitoring.tf:73]: added global here

File: modules/node-function-app/monitoring.tf [hunk 8]
```diff
@@ -1,130 +1,132 @@
...
+
+  action {
+    action_group_id = var.action_group_ids.iap
+  }
+}
+
+resource "azurerm_monitor_activity_log_alert" "function_app_delete" {
+  name                = "Function App Deleted - ${reverse(split("/", azurerm_linux_function_app.function_app.id))[0]}"
+  resource_group_name = var.resource_group_name
+  location            = "Global"
```
Comment (m-juckes-pins) [inline modules/node-function-app/monitoring.tf:105]: Global

These 2 changes are the only ones in the file

**Pull request closed without merging**